src/cat/cat.c

```diff
@@ -1,6 +1,325 @@
+    printf("Usage: cat [OPTION]... [FILE]...\n");                              \
+    printf("Concatenate FILE(s) to standard output.\n");                       \
+  } while (0)
+
+#define print_incorrect_args()                                                 \
+  do {                                                                         \
+    printf("incorrect arguments\n");                                           \
+    printf("see `cat --help`\n");                                              \
+  } while (0)
+
+#define BUF_MAX_LEN 4096 // max length of a buffer in bytes
+#define PATH_LEN 256     // max length of a path in bytes
+#define ARGS_MAX 16      // number of the max arguments
+#define ARGS_LEN 32      // max length of the arguments in bytes
+
+bool number_nonblank = false;  // number nonempty output lines, overrides -n
+bool show_ends = false;        // display $ at end of each line
+bool number = false;           // number all output lines
+bool squeeze_blank = false;    // suppress repeated empty output lines
+bool show_tabs = false;        // display TAB characters as ^I
+bool show_nonprinting = false; // use ^ and M- notation, except for LFD and TAB
+
+int cat(int filec, char **paths);
+int print_file(char *buf);
+int print_stdin(void);
+
+int main(int argc, char **argv) {
+  if (argc < 2) {
+    // print stdin
+    if (print_stdin() != 0) {
+      return 1;
+    }
+    return 0;
+  }
+
+  if (strcmp(argv[1], "--version") == 0) {
+    if (argc != 2) {
+      print_incorrect_args();
+      return 1;
+    }
+
+    print_version();
+    return 0;
+  }
+
+  if (strcmp(argv[1], "--help") == 0) {
+    if (argc != 2) {
+      print_incorrect_args();
+      return 1;
+    }
+
+    print_help();
+    return 0;
+  }
+
+  int filec = 0; // file count
+  char **paths = (char **)malloc(sizeof(char) * PATH_LEN * ARGS_MAX);
+  if (!paths) {
+    perror("could not allocate memory");
+    return 1;
+  }
+
+  // parse arguments
+  for (int i = 1; i < argc; ++i) {
+    int len = strlen(argv[i]);
+    if (len > 1 && argv[i][0] == '-') {
+      for (int j = 1; j < len; ++j) {
+        switch (argv[i][j]) {
+        case 'A':
+          show_nonprinting = true;
+          show_ends = true;
+          show_tabs = true;
+          continue;
+        case 'b':
+          number_nonblank = true;
+          number = false;
+          continue;
+        case 'e':
+          show_nonprinting = true;
+          show_ends = true;
+          continue;
+        case 'E':
+          show_ends = true;
+          continue;
+        case 'n':
+          number = true;
+          number_nonblank = false;
+          continue;
+        case 's':
+          squeeze_blank = true;
+          continue;
+        case 't':
+          show_nonprinting = true;
+          show_tabs = true;
+          continue;
+        case 'T':
+          show_tabs = true;
+          continue;
+        case 'v':
+          show_nonprinting = true;
+          continue;
```

Comment:
This can be simplified, see ls implementation

Reply:
```suggestion
      for (int j = 1; j < len; ++j) {
        switch (argv[i][j]) {
        case 'A':
          flags |= ShowNonprinting | ShowEnds | ShowTabs;
          break;
        case 'b':
          flags &= ~Number;
          flags |= NumberNonblank;
          break;
        case 'e':
          flags |= ShowNonprinting | ShowEnds;
          break;
        case 'n':
          flags &= ~NumberNonblank;
          flags |= Number;
          break;
        case 't':
          flags |= ShowNonprinting | ShowTabs;
          break;
        default:
          flags |= 1 << (stridx(FLAGLIST, argv[i][j]));
          break;
        }
      }
```